parsl/tests/test_monitoring/test_app_names.py

```diff
@@ -0,0 +1,99 @@
+"""Tests monitoring records app name under various decoration patterns.
+"""
+
+import logging
+import os
+import parsl
+import pytest
+import time
+
+from parsl.tests.configs.htex_local_alternate import fresh_config
+
+logger = logging.getLogger(__name__)
+
+
+@parsl.python_app
+def regular_decorated_app():
+    return 5
+
+
+@pytest.mark.local
+def get_regular_decorated_app():
+    return regular_decorated_app
+
+
+def for_decoration_later():
+    return 77
+
+
+def get_for_decoration_later():
+    return parsl.python_app(for_decoration_later)
+
+
+def get_decorated_closure():
+
+    r = 53
+
+    @parsl.python_app
+    def decorated_closure():
+        return r
+
+    return decorated_closure
+
+
+@pytest.mark.local
+@pytest.mark.parametrize("get_app,expected_name,expected_result",
+                         [(get_regular_decorated_app, "regular_decorated_app", 5),
+                          (get_for_decoration_later, "for_decoration_later", 77),
+                          (get_decorated_closure, "decorated_closure", 53)
+                          ])
+def test_app_name(get_app, expected_name, expected_result):
+
+    # this is imported here rather than at module level because
+    # it isn't available in a plain parsl install, so this module
+    # would otherwise fail to import and break even a basic test
+    # run.
+    import sqlalchemy
+
+    if os.path.exists("runinfo/monitoring.db"):
+        logger.info("Monitoring database already exists - deleting")
+        os.remove("runinfo/monitoring.db")
+
+    logger.info("Generating fresh config")
+    c = fresh_config()
+    logger.info("Loading parsl")
+    parsl.load(c)
+
+    logger.info("invoking and waiting for result")
+    app = get_app()
+    assert app().result() == expected_result
+
+    logger.info("cleaning up parsl")
+    parsl.dfk().cleanup()
+    parsl.clear()
+
+    logger.info("checking database content")
+    engine = sqlalchemy.create_engine("sqlite:///runinfo/monitoring.db")
```

Comment:
We're starting to use this "magic number" in a number of spots.  Perhaps not for this PR, but would be nice to reference a single definition somewhere.

Reply:
there's a more general factoring possibility I don't have my head round with the style of monitoring tests in general - tests I've written (not all committed) often look roughly like:

i) start some config that has monitoring
 - with a fresh monitoring db so that all thats in there is this test
ii) run a little bit of workflow - different per test
iii) shut down parsl - this is a barrier so that async monitoring information has "definitely" arrived in the DB
iv) make some assertions about that database, different per test (the motivation for starting with a clean DB is that because we started with a clean DB, we can be quite loose in queries, not having to select the particular workflow we just ran, for example)

There's a lot of boilerplate in there around the "run this workflow" "assert what monitoring saw" steps.